### DISCUSSION TITLE
Linear Algebra libraries for GPU

### DISCUSSION THREAD

**Discussion by dannys4** [MeasureTransport/MParT#235] [General]
Why we didn't use:
MAGMA: It imposed too stringent requirements on users to install dependencies, even though it provides everything we're looking for.
Kokkos kernels: Kokkos kernels provides mostly implements BLAS-like operations. However, it focuses on solving many small systems, and we need solutions to large systems.
Eigen: The GPU version is intended for small systems.

**Comment by mparno** [comment]
- Kokkos kernels provides BLAS functionality but it's direct solves are designed for solving many small independent systems. 
- Eigen can be compiled on the GPU, but it is meant for thread-level use and doesn't allow for GPU-acceleration of single systems.
- Magma provides everything we want and works on works on both NVIDIA and AMD gpus.    The [magma](https://github.com/MeasureTransport/MParT/tree/magma) branch provides an initial implementation that compiles but hangs for some unknown reason in `magma_queue_sync` (possibly a bug in magma).   It took some effort to automate the magma build within CMake and to link MParT with magma.   Magma can definitely be an option down the road, but I didn't have time to fully dig into this bug.
- CUBLAS and CUSOLVER are easy to install and link with MParT.  Together they also provide everything we need, but are part of the CUDA ecosystem and therefore do not provide support for non-NVIDIA GPUs.

**Reply by dannys4** [reply]
Additionally, we've talked about moving to HIPSolver, which should alleviate the portability concern at the moment while still enabling cusolver.
